controllers/control/cassandratask_controller_test.go

```diff
@@ -0,0 +1,278 @@
+	return func() {
+		By("Create Datacenter, pods and set dc status to Ready")
+		clusterName := fmt.Sprintf("test-%s", dcName)
+		testNamespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
+		Expect(k8sClient.Create(context.Background(), testNamespace)).Should(Succeed())
+
+		cassdcKey := types.NamespacedName{
+			Name:      dcName,
+			Namespace: namespace,
+		}
+
+		testDc := &cassdcapi.CassandraDatacenter{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        cassdcKey.Name,
+				Namespace:   cassdcKey.Namespace,
+				Annotations: map[string]string{},
+			},
+			Spec: cassdcapi.CassandraDatacenterSpec{
+				ClusterName:   clusterName,
+				ServerType:    "cassandra",
+				ServerVersion: "4.0.1",
+				Size:          3,
+			},
+			Status: cassdcapi.CassandraDatacenterStatus{},
+		}
+
+		Expect(k8sClient.Create(context.Background(), testDc)).Should(Succeed())
+
+		patchCassdc := client.MergeFrom(testDc.DeepCopy())
+		testDc.Status.CassandraOperatorProgress = cassdcapi.ProgressReady
+		testDc.Status.Conditions = []cassdcapi.DatacenterCondition{
+			{
+				Status: corev1.ConditionTrue,
+				Type:   cassdcapi.DatacenterReady,
+			},
+		}
+		Expect(k8sClient.Status().Patch(context.Background(), testDc, patchCassdc)).Should(Succeed())
+
+		for i := 0; i < int(testDc.Spec.Size); i++ {
+			pod := &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      fmt.Sprintf("test-cassdc-%s-pod%d", dcName, i),
+					Namespace: namespace,
+					Labels: map[string]string{
+						cassdcapi.ClusterLabel:    clusterName,
+						cassdcapi.DatacenterLabel: dcName,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "cassandra",
+							Image: "k8ssandra/cassandra-nothere:latest",
+						},
+					},
+				},
+			}
+			Expect(k8sClient.Create(context.Background(), pod)).Should(Succeed())
+
+			podIP := "127.0.0.1"
+
+			patchPod := client.MergeFrom(pod.DeepCopy())
+			pod.Status = corev1.PodStatus{
+				PodIP: podIP,
+				PodIPs: []corev1.PodIP{
+					{
+						IP: podIP,
+					},
+				},
+			}
+			Expect(k8sClient.Status().Patch(context.Background(), pod, patchPod)).Should(Succeed())
+		}
+	}
+}
+
+func createTask(command, namespace string) (types.NamespacedName, *api.CassandraTask) {
+	taskKey := types.NamespacedName{
+		Name:      fmt.Sprintf("test-%s-task", command),
+		Namespace: namespace,
+	}
+	task := &api.CassandraTask{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      taskKey.Name,
+			Namespace: taskKey.Namespace,
+		},
+		Spec: api.CassandraTaskSpec{
+			Datacenter: corev1.ObjectReference{
+				Name:      testDatacenterName,
+				Namespace: namespace,
+			},
+			Jobs: []api.CassandraJob{
+				{
+					Name:    fmt.Sprintf("%s-dc1", command),
+					Command: command,
+				},
+			},
+		},
+		Status: api.CassandraTaskStatus{},
+	}
+
+	return taskKey, task
+}
+
+var _ = Describe("Execute jobs against all pods", func() {
```

Comment:
Thanks for writing these tests. This is good stuff. There are probably more scenarios we could test, but they could make for good follow up PRs, particularly for someone else looking to get more familiar with the code.

Reply:
There are more scenarios, but writing them as envtest is difficult. Something like "prevent this from running concurrently" / "allow concurrently".